src/main/java/org/eclipse/uprotocol/communication/InMemorySubscriber.java

```diff
@@ -160,17 +155,20 @@ public CompletionStage<UStatus> unsubscribe(UUri topic, UListener listener, Call
 
         final UnsubscribeRequest unsubscribeRequest = UnsubscribeRequest.newBuilder().setTopic(topic).build();
         
-        return RpcMapper.mapResponseToResult(
+        CompletionStage<RpcResult<UnsubscribeResponse>> result = RpcMapper.mapResponseToResult(
             // Send the unsubscribe request
-            rpcClient.invokeMethod(UNSUBSCRIBE_METHOD, UPayload.pack(unsubscribeRequest), options), 
-            UnsubscribeResponse.class)
-            
+            rpcClient.invokeMethod(UNSUBSCRIBE_METHOD, UPayload.pack(unsubscribeRequest), options),     
+                UnsubscribeResponse.class)
             // Then unregister the listener
-            .thenCompose(response -> {
+            .whenComplete((response, ex) -> mHandlers.remove(topic));
```

Comment:
this means that you also remove the handler if we were not able to invoke USubscription::unsubscribe at all, e.g. because of a transport failure, right?

Reply:
+1, I think it would be better to remove the handler only when transport success.

Reply:
Yes, the handler will be removed regardless of whether the unsubscription is successful or not

Reply:
> +1, I think it would be better to remove the handler only when transport success.

I gave the same comment [here](https://github.com/eclipse-uprotocol/up-java/pull/141#discussion_r1669337211), but after discussing with Steven, we decided to ignore the transition from UNSUBSCRIBED_PENDING to UNSUBSCRIBED_STATE, so maintaining the handler is unnecessary.

Reply:
> I gave the same comment [here](https://github.com/eclipse-uprotocol/up-java/pull/141#discussion_r1669337211), but after discussing with Steven, we decided to ignore the transition from UNSUBSCRIBED_PENDING to UNSUBSCRIBED_STATE, so maintaining the handler is unnecessary.

Does that mean that once the state SUBSCRIBED has been reached, no additional notifications need to be forwarded to the client? If that is the case then we could automatically remove the handler once we get the SUBSCRIBED notification ...

Reply:
I want to keep the handler there incase we need to support other state changes in the future but right now the main use case is SUBSCRIBE_PENDING to SUBSCRIBED.